### PR TITLE
REFACTOR: Signup modal form validation

### LINKED ISSUE #4250: Validation messages for the login modal
- [ ] Login form validations are working fine for incorrect input fields(testing)
Unable to test, redirected to PL stable

### PR DESCRIPTION
**Pull request by VladimirMikulic**
[Demo](https://streamable.com/xevn9)

As the title suggests, I've added the login modal form validation.
I've also refactored form validation logic to be easily extendible for multiple types of forms.

Once the [PR for Webpack](https://github.com/publiclab/Leaflet.DistortableImage/pull/512) gets approved, I'll be happy to integrate Webpack with Rails in this project and refactor JavaScript to ESNext.

Resolves #4250

@jywarren could you review this? Thanks.

### CODE REVIEW

File: test/system/screenshots_test.rb [hunk 1]
```diff
@@ -71,6 +71,24 @@ class ScreenshotsTest < ApplicationSystemTestCase
     take_screenshot
   end
 
+  test 'login modal form validation' do
```
Comment (jywarren) [inline test/system/screenshots_test.rb]: Love the test!

**Comment by jywarren** [comment]
Hi, just noting that for login, the messages may need to be different; they still read as if you're inventing a new username and password:

image

Otherwise this looks really good! 

On a functional note, should we be validating on login, from a security perspective? Doesn't that help people to narrow their attempts to break in? I mean, a bit? Just want to bear this in mind as we consider this. cc @SidharthBansal and cc @tsparksh who has a keen eye for security!

**Comment by jywarren** [comment]
Yeah like, are we helping people find existing usernames? I can see how validation could stop people from entering their email address instead of their username, i guess, but some sites even allow that to be swapped anyways.

File: test/system/screenshots_test.rb (hunk 1 above)
Comment (VladimirMikulic) [reply test/system/screenshots_test.rb]: Thanks.

**Comment by VladimirMikulic** [comment]
@jywarren maybe instead of showing messages, we can just make inputs red?

**Comment by VladimirMikulic** [comment]
I can send ajax request to the server and if it doesn't return status between 200 and 300, I'll display a message like: 'Invalid username or password'.

**Comment by VladimirMikulic** [comment]
Apparently you don't have an API exposed to which I can send form data to and get a response.
You render error page or redirect to dashboard.

@SidharthBansal @jywarren there is no way to know whether the login form is valid or not on the client-side. I can check for username format or password format, but that's it.

In order to display the message 'Invalid username or password' I need an API, so I can send my data to it and get a proper response.

**Comment by Uzay-G** [comment]
Yeah I think you can only handle this part through Rails.

**Comment by SidharthBansal** [comment]
> I can check for username format or password format, but that's it.

We expect that only. Hence this is medium issue.

**Comment by SidharthBansal** [comment]
@Uzay-G kindly review
Travis is breaking. Kindly fix

**Comment by Uzay-G** [comment]
Yeah I will work on this when I am back at 3 :+1:

**Comment by SidharthBansal** [comment]
Sure.

**Comment by SidharthBansal** [comment]
@VladimirMikulic kindly make sure that all the issue and prs related to login-signup should have related milestone. If they don't please add it or ping me 
Thanks

**Comment by SidharthBansal** [comment]
@VladimirMikulic I will suggest you to break this in two prs. First pr where you are changing to meaningful names. Second pr for your changes. Thanks

**Comment by SidharthBansal** [comment]
It is difficult to review changes unrelated to your functionality here

**Comment by VladimirMikulic** [comment]
> For some reason, somehow this pull request is interfering with this system test:
> https://github.com/publiclab/plots2/blob/58d5c45682604179dc5953947a201b5c0de1f391/test/system/screenshots_test.rb#L18-L38
> 
> At line 34 Travis is saying that instead of getting the message "Please make sure password is at least 8 characters long with minimum one numeric value" it is getting "Passwords must be equal."

@Uzay-G I'll update it, leave it to me, I know why that is. Thank you for the investigation.

**Comment by SidharthBansal** [comment]
Try to fix this

**Comment by VladimirMikulic** [comment]
@SidharthBansal I've fixed it and renamed this PR properly.

**Comment by VladimirMikulic** [comment]
It should look like this. @Uzay-G your opinion on this matter?

Screenshot_20200112_170821

**Comment by SidharthBansal** [comment]
Travis is breaking and where is the second pr for variable name change?
I will merge that quickly before this one

**Comment by Uzay-G** [comment]
Hey! I think I might know why there is over and over again the `Log in` error for travis. @VladimirMikulic can you send me a screenshot of the login modal with your changes?

**Comment by VladimirMikulic** [comment]
@Uzay-G I appreciate your help, but this is no longer a login validation PR.
I am just refactoring sign up validation a bit as requested in #7200 

I will soon send a PR for login validation. Thanks a bunch!

**Comment by Uzay-G** [comment]
I understand @VladimirMikulic but somewhere in your changes for the signup validation something is interfering with the login UI. Could you send the screenshot? Thanks a lot :+1:

**Comment by VladimirMikulic** [comment]
@Uzay-G I remove all of login validation code.

**Comment by VladimirMikulic** [comment]
You have a demo video in the description. Hope it helps.

**Comment by SidharthBansal** [comment]
Once reviewed by @Uzay-G then ping me
thanks

**Comment by Uzay-G** [comment]
Hey @VladimirMikulic the code seems to be failing here, did you remove something important? There are many errors, could you direct me towards a specific issue? Then I can help you :+1:

**Comment by VladimirMikulic** [comment]
@Uzay-G WOW, I didn't expect that kind of support. Thank you very much.
I think that problem occurs because I rebased this PR and it collides with a lot of things.
There is no specific issue 😅
Please don't waste your time on this :smile: You've already done more than enough.

I'll debug it locally and once it passes Travis I'll ping you to check it out.

**Comment by Uzay-G** [comment]
> @Uzay-G WOW, I didn't expect that kind of support. Thank you very much.
> I think that problem occurs because I rebased this PR and it collides with a lot of things.
> There is no specific issue sweat_smile
> Please don't waste your time on this smile You've already done more than enough.
> 
> I'll debug it locally and once it passes Travis I'll ping you to check it out.

I'm happy to help out! If you encounter a problem just ping me and I'll see if I can do anything. :+1:

**Pull request closed without merging**